command/operator_client_state.go

```diff
@@ -0,0 +1,117 @@
+package command
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/hashicorp/go-hclog"
+	trstate "github.com/hashicorp/nomad/client/allocrunner/taskrunner/state"
+	"github.com/hashicorp/nomad/client/state"
+	"github.com/posener/complete"
+)
+
+type OperatorClientStateCommand struct {
+	Meta
+}
+
+func (c *OperatorClientStateCommand) Help() string {
+	helpText := `
+Usage: nomad operator client-state <path_to_nomad_dir>
+  Emits a json representation of the stored client state in json form.
+`
+	return strings.TrimSpace(helpText)
+}
+func (c *OperatorClientStateCommand) AutocompleteFlags() complete.Flags {
+	return complete.Flags{}
+}
+
+func (c *OperatorClientStateCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictNothing
+}
+
+func (c *OperatorClientStateCommand) Synopsis() string {
+	return "Dump the nomad client state"
+}
+func (c *OperatorClientStateCommand) Name() string { return "operator client-state" }
+
+func (c *OperatorClientStateCommand) Run(args []string) int {
+	if len(args) != 1 {
+		c.Ui.Error("This command takes one argument: <nomad-data-dir>")
+		c.Ui.Error(commandErrorText(c))
+
+		return 1
+	}
+
+	logger := hclog.L()
+	db, err := state.NewBoltStateDB(logger, args[0])
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("failed to open client state: %v\n", err))
+		return 1
+	}
+	defer db.Close()
+
+	allocs, _, err := db.GetAllAllocations()
+	if err != nil {
+		c.Ui.Error(fmt.Sprintf("failed to get allocations: %v\n", err))
+		return 1
+	}
+
+	data := map[string]*clientStateAlloc{}
```

Comment:
Right now this command is only showing the client state for the allocations bucket. But if you take a look at the [client state schema](https://github.com/hashicorp/nomad/blob/v1.4.3/client/state/db_bolt.go#L24-L46) there's also data for plugins, drivers, and devices. It's ok that we don't include that in this PR (we can always come back and add support for that later), but we should probably make sure that our output has a top-level Allocations key for extensibility. That way our output is something like:

```json
{
  "Allocations": [
    "694dabb6-2877-4093-8f1b-7f8d9d8c5d48": {
      "Alloc": {...},
      "DeployStatus": {...},
      "Tasks": {...},
    }
  ]
}
```

And in a future PR we can extend it to be something like:

```json
{
  "Allocations": [
    "694dabb6-2877-4093-8f1b-7f8d9d8c5d48": {
      "Alloc": {...},
      "DeployStatus": {...},
      "Tasks": {...},
    }
  ],
  "Devices": [...],
  "Drivers": [...],
  "DynamicPlugins": [...],
}

```